pkg/audit/audit.go

```diff
@@ -0,0 +1,106 @@
+// Copyright (C) 2018 Storj Labs, Inc.
+// See LICENSE for copying information.
+
+package audit
+
+import (
+	"bytes"
+	"context"
+	"io"
+
+	"github.com/vivint/infectious"
+
+	"storj.io/storj/pkg/pb"
+	"storj.io/storj/pkg/piecestore/rpc/client"
+)
+
+// Auditor struct
+type Auditor struct {
+	ps client.PSClient
+}
+
+// NewAuditor creates a new instance of an Auditor struct
+func NewAuditor(ps client.PSClient) *Auditor {
+	return &Auditor{ps: ps}
+}
+
+func (a *Auditor) getShare(ctx context.Context, stripeIndex, shareSize, stripeNumber int,
+	id client.PieceID, pieceSize int64) (share infectious.Share, err error) {
+
+	rr, err := a.ps.Get(ctx, id, pieceSize, &pb.PayerBandwidthAllocation{})
+	if err != nil {
+		return infectious.Share{}, err
+	}
+
+	offset := shareSize * stripeIndex
+
+	rc, err := rr.Range(ctx, int64(offset), int64(shareSize))
+	if err != nil {
+		return infectious.Share{}, err
+	}
+
+	buf := make([]byte, shareSize)
+	_, err = io.ReadFull(rc, buf)
+	if err != io.EOF || err != io.ErrUnexpectedEOF {
+		return infectious.Share{}, err
+	}
+
+	share = infectious.Share{
+		Number: stripeNumber,
+		Data:   buf,
+	}
+	return share, nil
+}
+
+func (a *Auditor) audit(ctx context.Context, required, total int, originals []infectious.Share) (badStripes []int, err error) {
+	f, err := infectious.NewFEC(required, total)
+	if err != nil {
+		return nil, err
+	}
+	copies := make([]infectious.Share, len(originals))
+	for _, copy := range originals {
+		copies[copy.Number] = copy.DeepCopy()
+	}
+	err = f.Correct(copies)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, share := range copies {
+		if !bytes.Equal(originals[share.Number].Data, share.Data) {
```

Comment:
If I am not wrong, there is not guarantee that `originals` contains the full house of total (N) shares and they are sorted by the `share.Number`. Hence, `original[share.Number]` is error prone. I should be rather `originals[i]` where `i` comes from `for i, share := range copies`.

Similar comment for copying `originals` to `copies` above.